src/main/java/org/craftercms/studio/impl/v2/service/security/AuthenticationChainImpl.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2007-2018 Crafter Software Corporation. All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package org.craftercms.studio.impl.v2.service.security;
+
+import org.craftercms.studio.api.v1.service.activity.ActivityService;
+import org.craftercms.studio.api.v1.util.StudioConfiguration;
+import org.craftercms.studio.api.v2.dal.GroupDAO;
+import org.craftercms.studio.api.v2.dal.UserDAO;
+import org.craftercms.studio.api.v2.service.security.AuthenticationChain;
+import org.craftercms.studio.api.v2.service.security.AuthenticationProvider;
+import org.craftercms.studio.api.v2.service.security.internal.UserServiceInternal;
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.core.io.Resource;
+import org.yaml.snakeyaml.Yaml;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import static org.craftercms.engine.targeting.impl.TargetedContentStoreAdapter.logger;
+
+public class AuthenticationChainImpl implements AuthenticationChain {
+
+    private List<AuthenticationProvider> authentitcationChain;
+
+    private UserServiceInternal userServiceInternal;
+    private ActivityService activityService;
+    private StudioConfiguration studioConfiguration;
+    private UserDAO userDao;
+    private GroupDAO groupDao;
+
+    public void init() {
+        Resource resource = new ClassPathResource(studioConfiguration.getProperty(
+                StudioConfiguration.CONFIGURATION_AUTHENTICATION_CHAIN_CONFIG_LOCATION));
+        try (InputStream in = resource.getInputStream()) {
+            Yaml yaml = new Yaml();
```

Comment:
To specify the authentication chain, can't we just use the same studio config property `studio.security.type` and separate the chain by commas, eg. `studio.security.type: headers, ldap, db`?

Reply:
Security type will be removed in 3.1.1 (https://github.com/craftercms/craftercms/issues/2389)